File: Week_08-位运算-布隆过滤-排序/56.合并区间.cpp
/*
 * @lc app=leetcode.cn id=56 lang=cpp
 *
 * [56] 合并区间
 */
/*
先对未合并区间按照左端点大小排序；
第一个未合并区间放入合并区间数组；
遍历剩余未合并区间，
如果待合并区间左端点大于已合并区间数组最后一个区间的右端点， 说明这是一个非连续区间，push到合并区间数组；
如果待合并区间左端点小于已合并区间数组的最后一个区间的右端点，则把此合并区间数组的最后一个区间更新为（原右端点和未合并区间右端点的最大值）
*/

// @lc code=start
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.size() == 0) return {};
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> merged;
        for (int i=0; i<intervals.size(); i++) {
            int L = intervals[i][0], R = intervals[i][1];
            if (!merged.size() || merged.back()[1] < L) merged.push_back({L, R});
            else merged.back()[1] = max(merged.back()[1], R);
        }
        return merged;
    }
};
// @lc code=end

File: Week_08-位运算-布隆过滤-排序/56.合并区间.py
#
# @lc app=leetcode.cn id=56 lang=python
#
# [56] 合并区间
#
'''
* 首先，将列表中的区间按照左端点升序排序。然后将第一个区间加入 merged 数组中，并按顺序依次考虑之后的每个区间：
* 如果当前区间的左端点在数组 merged 中最后一个区间的右端点之后，那么它们不会重合，可以直接将这个区间加入数组 merged 的末尾；
* 否则，它们重合，需要用当前区间的右端点更新数组 merged 中最后一个区间的右端点，将其置为二者的较大值。
'''

# @lc code=start
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        intervals.sort(key = lambda x: x[0])  #数组排序
        merged = []
        for interval in intervals:
            # 如果列表为空，或者当前区间与上一区间不重合，直接添加
            if not merged or merged[-1][1] < interval[0]:
                merged.append(interval)
            else:
            # 否则的话，就可以与上一区间进行合并
                merged[-1][1] = max(merged[-1][1], interval[1])
        return merged
# @lc code=end

